components/style/traversal.rs

```diff
@@ -141,12 +141,23 @@ pub fn remove_from_bloom_filter<'a, N, C>(context: &C, root: OpaqueNode, node: N
 
 pub trait DomTraversalContext<N: TNode>  {
     type SharedContext: Sync + 'static;
+
     fn new<'a>(&'a Self::SharedContext, OpaqueNode) -> Self;
+
     /// Process `node` on the way down, before its children have been processed.
-    fn process_preorder(&self, node: N);
+    fn process_preorder(&self, node: N) -> ForceTraversalStop;
+
     /// Process `node` on the way up, after its children have been processed.
+    ///
+    /// This is only executed if `has_postorder_traversal` returns true.
     fn process_postorder(&self, node: N);
 
+    /// Boolean that specifies whether a bottom up traversal should be
+    /// performed.
+    ///
+    /// If it's false, then process_postorder has no effect at all.
+    fn has_postorder_traversal(&self) -> bool { true }
```

Comment:
nit: How about `needs_postorder_traversal`? Every traversal *has* a postorder traversal; it's just whether the traversal *needs* to be done that varies.